### PR TITLE
fix(client): prefer CLI arguments, following OpenSSH behaviour

### PR DESCRIPTION
**Pull request by TheoTechnicguy**
This PR fixes issue #35 where the client preferred the values from the configuration file `~/.ssh/config` instead of the CLI arguments. The new behaviour reflects the one used by OpenSSH, preferring the port from the CLI and the hostname from the configuration.

### CODE REVIEW

**Comment by francoismichel** [comment]
The integration test suite might be a bit agressive sometimes, I'll run it locally later and let you know

**Comment by TheoTechnicguy** [comment]
How do I run the tests? `go test ./...` says everything is OK.

**Comment by TheoTechnicguy** [comment]
In my opinion, the failures are not a result of code changes, as I get the same errors when running integration tests on the main branch.

[asciicast](https://asciinema.org/a/nQQ1oc7kwZUsyuB7sBm231NuZ)

I'm not familiar with the Ginko test framework, so I don't know where the faults are coming from. I am willing to help resolve the failing tests with time assistance.
Furthermore, I find it interesting that macOS builds pass, but Linux do not. Do you happen to know the reason for this?

**Comment by francoismichel** [comment]
Sorry for the delay, I am bit busy on week ends :-)

Sooo, it is a bit surprising, especially that I am not able to reproduce the fails on my own setup on several distros. It seems either the server in the integration tests crashes at some point. What is also weird is that I cannot reproduce the fails on the main branch when I rerun the Github action on it, but the error seems unrelated to your PR. We probably need a bit more logging on the server-side, and in the test suite, check that the server [`Consistently`](https://onsi.github.io/gomega/#consistently) does not crash with an error, so that we can get the actual error if it crashes. I can try to add that additional checks on the test suite to debug that.

Thanks for the PR ad the help ! Let's continue to investigate.

**Comment by francoismichel** [comment]
Allright, both the integration test suite and your PR seem fine, the problem is probably the github secrets that do not seem accessible in PRs from non-maintainers. These secrets are not real secrets so I'll remove these and generate all the stuff (mock certs and keys) in the CI script. Let's get back to the real topic of this PR, I'll do a review soon. :-)

**Comment by TheoTechnicguy** [comment]
Thank you for investing the time looking into this. I have just merged with `main`. We'll see how GitHub likes the changes :yum:

**Comment by francoismichel** [comment]
Github seems to love it.
I'll do the PR review tomorrow.
Thanks gain!

**Comment by TheoTechnicguy** [comment]
Nice! Your changes seem to make GitHub happy, which makes me happy (^u^). Thanks again for investigating the problem. Please don't hesitate to ask for changes if needed.

File: cli/client/main.go [hunk 1]
```diff
@@ -416,14 +413,33 @@ func mainWithStatusCode() int {
...
+	// |    yes   |  ok   |     yes     | use CLI port         |
+	// |    yes   |  ok   |     no      | use CLI port         |
+	// |    no    |  ---  |     yes     | use config port      |
+	// |    no    |  ---  |     no      | use default port     |
+
+	var port int
+	if urlPort != "" {
+		if parsedPort, err := strconv.Atoi(urlPort); err != nil || parsedPort > 0xffff {
+			// use WithLevel(zerolog.FatalLevel) to log a fatal level, but let us handle
+			// programm termination. log.Fatal() exits with os.Exit(1).
```
Comment (glensc) [inline cli/client/main.go:432]:
```suggestion
			// program termination. log.Fatal() exits with os.Exit(1).
```

Comment (TheoTechnicguy) [reply cli/client/main.go]: Nice catch!

File: cli/client/main.go [hunk 2]
```diff
@@ -416,14 +413,33 @@ func mainWithStatusCode() int {
 	}
 
 	hostnameIsAnIP := net.ParseIP(hostname) != nil
-
-	port := configPort
-	if port == -1 && urlPort != "" {
-		port, err = strconv.Atoi(urlPort)
-		if err != nil {
-			log.Error().Msgf("invalid port number: %s: %s", urlPort, err)
+	// *** Port selection ***
```
Comment (francoismichel) [inline cli/client/main.go:416]: Overall I am not a huge fan of the table. I think we can make it more concise. The error case when the port cannot be parsed seems obvious.
I think it is fine if we just say that the CLI port takes precedence over the config port when specified.

File: cli/client/main.go [hunk 3]
```diff
@@ -416,14 +413,33 @@ func mainWithStatusCode() int {
...
+	// |    yes   |  nok  |     no      | throw error and exit |
+	// |    yes   |  ok   |     yes     | use CLI port         |
+	// |    yes   |  ok   |     no      | use CLI port         |
+	// |    no    |  ---  |     yes     | use config port      |
+	// |    no    |  ---  |     no      | use default port     |
+
+	var port int
+	if urlPort != "" {
+		if parsedPort, err := strconv.Atoi(urlPort); err != nil || parsedPort > 0xffff {
+			// use WithLevel(zerolog.FatalLevel) to log a fatal level, but let us handle
```
Comment (francoismichel) [inline cli/client/main.go:431]: Would you mind inverting the two parts if this branch ? Putting the non-error case in the `if` part and the error case in the `else` part ? I think it makes it more readable because the default flow goes towards what we want to achieve. But maybe it is just me ? If you have good reasons to let it like that, let me know.

File: cli/client/main.go [hunk 4]
```diff
@@ -416,14 +413,33 @@ func mainWithStatusCode() int {
...
+			// program termination. log.Fatal() exits with os.Exit(1).
+			log.WithLevel(zerolog.FatalLevel).Str("Port", urlPort).Err(err).Msg("cli contains an invalid port")
+			fmt.Fprintf(os.Stderr, "Bad port '%s'\n", urlPort)
 			return -1
+		} else {
+			port = parsedPort
 		}
+	} else if configPort != -1 {
```
Comment (francoismichel) [inline cli/client/main.go:439]: Add the comment `// no port specified in the CLI args` ?

File: cli/client/main.go [hunk 5]
```diff
@@ -416,14 +413,33 @@ func mainWithStatusCode() int {
...
+		port = configPort
+	} else {
+		port = 443
```
Comment (francoismichel) [inline cli/client/main.go:442]: Add a comment, like `// no explicit port from the CLI and no port specified in config` ?

**Review by francoismichel — CHANGES_REQUESTED** [review]
Thank you for the PR !

Just a few comments but it's great. :-)

If you disagree with the comments we can iterate.

File: cli/client/main.go (hunk 3 above)
Comment (TheoTechnicguy) [reply cli/client/main.go:431]: Sure, will do. I think this is more a go habit, as most of my `if` statements are used to check for the absence of errors :smile:

File: cli/client/main.go [hunk 6]
```diff
@@ -556,16 +543,16 @@ func mainWithStatusCode() int {
...
 	if err != nil {
-		if transportErr, ok :=  err.(*quic.TransportError); ok {
+		if transportErr, ok := err.(*quic.TransportError); ok {
 			if transportErr.ErrorCode.IsCryptoError() {
 				if tty == nil {
 					log.Error().Msgf("insecure server cert in non-terminal session, aborting")
 					return -1
 				}
 				if _, ok = knownHosts[hostname]; ok {
-					log.Error().Msgf("The server certificate cannot be verified using the one installed in %s. " +
```
Comment (francoismichel) [inline cli/client/main.go:550]: I am okay to reformat the code in a standard manner but could we do that in another PR ? :-)

**Review by francoismichel — CHANGES_REQUESTED** [review]
Great ! If you could just remove the part when the code is reformatted it it would make it easier to review and make the final commit cleaner. :-)

Other than that, I am happy with your changes !

File: cli/client/main.go (hunk 6 above)
Comment (TheoTechnicguy) [reply cli/client/main.go:550]: Ouch! That was not intended — you're absolutely right about keeping one topic per PR.

Comment (TheoTechnicguy) [reply cli/client/main.go]: Done! I don't want to monopolize your time, so I won't re-re-request a review at this time.

Comment (francoismichel) [reply cli/client/main.go]: You can request it anytime, don't worry.

**Review by francoismichel — APPROVED** [review]
Looks good, thanks!
